src/pkg/p2p/preheat/dao/instance/dao.go

```diff
@@ -0,0 +1,149 @@
+package instance
+
+import (
+	"context"
+
+	beego_orm "github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/orm"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/pkg/p2p/preheat/models/provider"
+)
+
+// DAO for instance
+type DAO interface {
+	Create(ctx context.Context, instance *provider.Instance) (int64, error)
+	Get(ctx context.Context, id int64) (*provider.Instance, error)
+	Update(ctx context.Context, instance *provider.Instance, props ...string) error
+	Delete(ctx context.Context, id int64) error
+	Count(ctx context.Context, query *q.Query) (total int64, err error)
+	List(ctx context.Context, query *q.Query) (ins []*provider.Instance, err error)
+}
+
+// New instance dao
+func New() DAO {
+	return &dao{}
+}
+
+// ListInstanceQuery defines the query params of the instance record.
+type ListInstanceQuery struct {
+	Page     uint
+	PageSize uint
+	Keyword  string
+}
+
+type dao struct{}
+
+var _ DAO = (*dao)(nil)
+
+// Create adds a new distribution instance.
+func (d *dao) Create(ctx context.Context, instance *provider.Instance) (int64, error) {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return 0, err
+	}
+	return o.Insert(instance)
+}
+
+// Get gets instance from db by id.
+func (d *dao) Get(ctx context.Context, id int64) (*provider.Instance, error) {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	di := provider.Instance{ID: id}
+	err = o.Read(&di, "ID")
+	if err == beego_orm.ErrNoRows {
+		return nil, nil
+	}
+	return &di, err
+}
+
+// Update updates distribution instance.
+func (d *dao) Update(ctx context.Context, instance *provider.Instance, props ...string) error {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+	err = o.Begin()
+	if err != nil {
+		return err
+	}
+
+	// check default instances first
```

Comment:
default handler should happen at least in the mgr or controller layer, not DAO here. DAO only needs to handle the concrete SQL actions.